nano/node/request_aggregator.cpp

```diff
@@ -71,104 +76,115 @@ void nano::request_aggregator::stop ()
 
 std::size_t nano::request_aggregator::size () const
 {
-	nano::unique_lock<nano::mutex> lock{ mutex };
-	return requests.size ();
+	nano::lock_guard<nano::mutex> lock{ mutex };
+	return queue.size ();
 }
 
 bool nano::request_aggregator::empty () const
 {
-	return size () == 0;
+	nano::lock_guard<nano::mutex> lock{ mutex };
+	return queue.empty ();
 }
 
-// TODO: This is badly implemented, will prematurely drop large vote requests
-void nano::request_aggregator::add (std::shared_ptr<nano::transport::channel> const & channel_a, std::vector<std::pair<nano::block_hash, nano::root>> const & hashes_roots_a)
+bool nano::request_aggregator::request (request_type const & request, std::shared_ptr<nano::transport::channel> const & channel)
 {
+	release_assert (channel != nullptr);
+
+	// This should be checked before calling request
 	debug_assert (wallets.reps ().voting > 0);
-	bool error = true;
-	auto const endpoint (nano::transport::map_endpoint_to_v6 (channel_a->get_endpoint ()));
+	debug_assert (!request.empty ());
+
+	bool added = false;
+	{
+		nano::lock_guard<nano::mutex> guard{ mutex };
+		added = queue.push ({ request, channel }, { nano::no_value{}, channel });
+	}
+	if (added)
+	{
+		stats.inc (nano::stat::type::request_aggregator, nano::stat::detail::request);
+		stats.add (nano::stat::type::request_aggregator, nano::stat::detail::request_hashes, request.size ());
+
+		condition.notify_one ();
+	}
+	else
+	{
+		stats.inc (nano::stat::type::request_aggregator, nano::stat::detail::overfill);
+		stats.add (nano::stat::type::request_aggregator, nano::stat::detail::overfill_hashes, request.size ());
+	}
+	return added;
+}
+
+void nano::request_aggregator::run ()
+{
 	nano::unique_lock<nano::mutex> lock{ mutex };
-	// Protecting from ever-increasing memory usage when request are consumed slower than generated
-	// Reject request if the oldest request has not yet been processed after its deadline + a modest margin
-	if (requests.empty () || (requests.get<tag_deadline> ().begin ()->deadline + 2 * this->max_delay > std::chrono::steady_clock::now ()))
+	while (!stopped)
 	{
-		auto & requests_by_endpoint (requests.get<tag_endpoint> ());
-		auto existing (requests_by_endpoint.find (endpoint));
-		if (existing == requests_by_endpoint.end ())
+		stats.inc (nano::stat::type::request_aggregator, nano::stat::detail::loop);
+
+		if (!queue.empty ())
 		{
-			existing = requests_by_endpoint.emplace (channel_a).first;
+			run_batch (lock);
+			debug_assert (!lock.owns_lock ());
+			lock.lock ();
```

Comment:
Perhaps run_batch should be responsible for re-locking i.e. returning the lock to the state it was when passed in.

Reply:
I used this pattern in other classes and didn't want to unnecessarily change it. Does it matter that much as long as the post condition is always met?